pkg/operator/util/ca.go

```diff
@@ -0,0 +1,71 @@
+package util
+
+import (
+	crand "crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"time"
+
+	cryptutil "github.com/openshift/library-go/pkg/crypto"
+)
+
+const SigningCertificateLifetimeInDays = 365 // 1 year
+
+// CertHalfwayExpired indicates whether half of the cert validity period has elapsed.
+func CertHalfwayExpired(cert *x509.Certificate) bool {
+	halfValidPeriod := cert.NotAfter.Sub(cert.NotBefore).Nanoseconds() / 2
+	halfExpiration := cert.NotBefore.Add(time.Duration(halfValidPeriod) * time.Nanosecond)
+	return time.Now().After(halfExpiration)
+}
+
+// RotateSigningCA creates a new CA with a set of cross-signed interim CAs to allow for graceful rollover.
+// Returns new CA pem, new CA key pem and the full client CA bundle.
+func RotateSigningCA(currentCert *x509.Certificate, currentKey *rsa.PrivateKey) ([]byte, []byte, []byte, error) {
+	// Generate a new signing cert
+	newCertConfig, err := cryptutil.MakeSelfSignedCAConfigForSubject(currentCert.Subject, SigningCertificateLifetimeInDays)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	newCert := newCertConfig.Certs[0]
+	newKey := newCertConfig.Key
+
+	// The first interim CA comprises the current CA's public key, private key, and subject. It's self-issued but not
+	// self-signed as it's signed by the new CA key. This creates a trust bridge between refreshed clients and
+	// unrefreshed servers.
+	rawSignedByNewCert, err := x509.CreateCertificate(crand.Reader, currentCert, currentCert, currentCert.PublicKey, newKey)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	signedByNewCert := &x509.Certificate{Raw: rawSignedByNewCert}
+
+	// The second interim CA comprises the new CA's public key, private key, and subject. It's self-issued but not
+	// self-signed as it's signed by the current CA key. This creates a trust bridge between the unrefreshed clients and
+	// refreshed servers as long as refreshed servers serve with a bundle containing this CA and the serving cert.
+	rawSignedByCurrentCert, err := x509.CreateCertificate(crand.Reader, newCert, newCert, newCert.PublicKey, currentKey)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	signedByCurrentCert := &x509.Certificate{Raw: rawSignedByCurrentCert}
+
+	// Create the full CA bundle comprising all 4 certs
+	// TODO(marun) Is the ordering of certs in the bundle important?
+	bundleCerts := []*x509.Certificate{
+		currentCert,
+		signedByNewCert,
```

Comment:
This cert may not even be needed since we have the old CA in the trust bundle. It would be nice to have an e2e test which proves this.

Reply:
I think the [design doc](https://docs.google.com/document/d/1VqrIERs30M9EyPaqcO43wcSXHXvfe_Ao2W0ThDIohlM/edit) was presuming a signing scheme that relied on intermediates, but since that's not the strategy used by this operator (all service certs are signed by the current 'root'), the only intermediate that matters is #4 (including it with service certs supports support clients that haven't yet updated to the newest ca bundle). That implies that neither intermediate cert needs to be part of the bundle since the old and new certs (`#1` and `#2`) are by themselves sufficient to validate `X` and `Y`. `#3` is unnecessary.